src/operation/base.hpp

```diff
@@ -19,35 +19,79 @@ namespace ddwaf::operation {
+    base(base &&) noexcept = default;
     base &operator=(const base &) = default;
-    base &operator=(base &&) = default;
+    base &operator=(base &&) noexcept = default;
+
+    [[nodiscard]] virtual DDWAF_OBJ_TYPE supported_type() const = 0;
+    // Returns a string representing this particular instance of the operator, for example,
+    // an operator matching regexes could provide the regex as its string representation.
+    [[nodiscard]] virtual std::string_view to_string() const = 0;
+    // The return value of this function should outlive the function scope,
+    // for example, through a constexpr class static string_view initialised
+    // with a literal.
+    [[nodiscard]] virtual std::string_view name() const = 0;
+
+    [[nodiscard]] virtual std::pair<bool, memory::string> match(const ddwaf_object &obj) const = 0;
+};
+
+template <typename T> class base_impl : public base {
+public:
+    using ptr = std::shared_ptr<base>;
 
-    [[nodiscard]] virtual std::optional<event::match> match(std::string_view str) const = 0;
+    base_impl() = default;
+    ~base_impl() override = default;
+    base_impl(const base_impl &) = default;
+    base_impl(base_impl &&) noexcept = default;
+    base_impl &operator=(const base_impl &) = default;
+    base_impl &operator=(base_impl &&) noexcept = default;
 
-    [[nodiscard]] virtual std::optional<event::match> match_object(const ddwaf_object &obj) const
+    [[nodiscard]] std::string_view to_string() const override
     {
-        if (obj.stringValue == nullptr) {
-            return std::nullopt;
-        }
-        return match({obj.stringValue, static_cast<std::size_t>(obj.nbEntries)});
+        return static_cast<const T *>(this)->to_string_impl();
     }
 
-    [[nodiscard]] virtual std::string_view to_string() const { return ""; }
+    [[nodiscard]] std::string_view name() const override { return T::name_impl(); }
 
-    /* The return value of this function should outlive the function scope,
-     * for example, through a constexpr class static string_view initialised
-     * with a literal. */
-    [[nodiscard]] virtual std::string_view name() const = 0;
+    [[nodiscard]] DDWAF_OBJ_TYPE supported_type() const override
+    {
+        return T::supported_type_impl();
+    }
 
-    [[nodiscard]] event::match make_event(std::string_view resolved, std::string_view matched) const
+    // Helper used for testing purposes
+    template <typename U> [[nodiscard]] std::pair<bool, memory::string> match(const U &data) const
     {
-        return {memory::string(resolved), memory::string(matched), name(), to_string(), {}, {}};
+        return static_cast<const T *>(this)->match_impl(data);
+    }
+
+    [[nodiscard]] std::pair<bool, memory::string> match(const ddwaf_object &obj) const override
+    {
+        const auto *ptr = static_cast<const T *>(this);
+        if constexpr (T::supported_type_impl() == DDWAF_OBJ_STRING) {
+            if (obj.stringValue == nullptr) {
```

Comment:
you're looking at the union value without checking the union type tag. idk if this is checked by the caller, but even if it is, this being a public a method it shouldn't be unsafe

Reply:
Right, I'll add a check.